libsrc/vic20/tgi/vic20-hi.s

```diff
@@ -0,0 +1,580 @@
+        .byte >(VBASE + YRES * 8)
+        .byte >(VBASE + YRES * 9)
+        .byte >(VBASE + YRES * 10)
+        .byte >(VBASE + YRES * 11)
+        .byte >(VBASE + YRES * 12)
+        .byte >(VBASE + YRES * 13)
+        .byte >(VBASE + YRES * 14)
+        .byte >(VBASE + YRES * 15)
+        .byte >(VBASE + YRES * 16)
+        .byte >(VBASE + YRES * 17)
+        .byte >(VBASE + YRES * 18)
+        .byte >(VBASE + YRES * 19)
+
+.code
+
+; ------------------------------------------------------------------------
+; INSTALL routine. Is called after the driver is loaded into memory. May
+; initialize anything that has to be done just once. Is probably empty
+; most of the time.
+;
+; Must set an error code: NO
+;
+
+INSTALL:
+        rts
+
+
+; ------------------------------------------------------------------------
+; UNINSTALL routine. Is called before the driver is removed from memory. May
+; clean up anything done by INSTALL but is probably empty most of the time.
+;
+; Must set an error code: NO
+;
+
+UNINSTALL:
+        rts
+
+
+; ------------------------------------------------------------------------
+; INIT: Changes an already installed device from text mode to graphics
+; mode.
+; Note that INIT/DONE may be called multiple times while the driver
+; is loaded, while INSTALL is only called once, so any code that is needed
+; to initializes variables and so on must go here. Setting palette and
+; clearing the screen is not needed because this is called by the graphics
+; kernel later.
+; The graphics kernel will never call INIT when a graphics mode is already
+; active, so there is no need to protect against that.
+;
+; Must set an error code: YES
+;
+
+INIT:
+
+; Initialize variables
+
+        ldx     #$FF
+        stx     BITMASK
+
+; Make screen columns.
+        lda     #<SBASE
+        sta     tmp2
+        lda     #>SBASE
+        sta     tmp2+1
+        ldx     #0
+l:      ldy     #0
+        txa
+        clc
+        adc     #$10
+m:      sta     (tmp2),y
+        clc
+        adc     #12
+        iny
+        cpy     #20
+        bne     m
+        lda     tmp2
+        clc
+        adc     #20
+        sta     tmp2
+        bcc     n
+        inc     tmp2+1
+n:      inx
+        cpx     #12
+        bne     l
+
+; Set up VIC.
+        ldx #5
+l2:     clc
+        lda     $ede4,x
+        adc     VICREGS,x
+        sta     $9000,x
+        dex
+        bpl     l2
+        lda     $900f
+        and     #%00000111
+        ora     #8
+        sta     $900f
+
+; Done, reset the error code
+
+        lda     #TGI_ERR_OK
+        sta     ERROR
+        rts
+
+; ------------------------------------------------------------------------
+; DONE: Will be called to switch the graphics device back into text mode.
+; The graphics kernel will never call DONE when no graphics mode is active,
+; so there is no need to protect against that.
+;
+; Must set an error code: NO
+;
+
+;DONE:   jmp    $e518       ; KERNAL VIC init.
```

Comment:
* I have a pretty strong opinion against commented-out code without a reason saying why the code is commented-out.
* I'd rather spend the 3 bytes and have all entries in the jump table actually end up in the driver code.